theano/compile/builders.py

```diff
@@ -1,44 +1,79 @@
 from __future__ import absolute_import, print_function, division
+from functools import reduce
+
 import theano
 from theano import gof
 from theano.compat import izip
 from theano.compile.function_module import orig_function
-from theano.compile import SharedVariable, rebuild_collect_shared
+from theano.compile import SharedVariable, rebuild_collect_shared, optdb
 from theano.gof import ops_with_inner_function
 from theano.gof.graph import io_connection_pattern
-
-from functools import reduce
+from theano.gof.utils import undef
 
 
 class OpFromGraph(gof.Op):
     """
     This creates an `Op` from inputs and outputs lists of variables.
-
     The signature is similar to theano.function() and the resulting
     `Op`'s perform will do the same operation as::
 
         orig_function(inputs, outputs, **kwargs)
+    Currently does not support 'updates' or 'givens' argument.
+
+    Parameters
+    ----------
+
+    inputs: list of variables
+    outputs: list of variables
+    inline: bool, optional
+        if True, will cause the Op's original graph being used during
+        compilation, otherwise will use a pre-compiled function inside.
+    grad_overrides: None | undef | OpFromGraph instance | function | \
+        list of (None|undef|function), optional
+        Used to override default gradient routine.
+        Overriding function(s) must take two list of variable(s) as inputs,
+        the original inputs and ups gradients
+        For different `grad_overrides`:
+
+        - `None` : will use default gradient routine.
+        - theano.utils.undef : No gradient will be used (zero)
+        - OpFromGraph instance: the OfG instance should accept inputs with same
+            order and types as specified in "inputs" and "outputs" arguments
+        - function : must return list of Variable.
+        - list : each function must return a single Variable. The order
+            of the list must corresponds to inputs
+    rop_overrides: None | undef | OpFromGraph instance | function | \
+        list of (None|undef|function), optional
+        similar to grad_overrides, list order should match two list of "inputs"
+        concatenated.
 
     TODO:
         - examples for a multi-layer mlp. where?
         - __hash__, __eq__ otherwise won't merge, try
-          gof.opt.is_same_graph_with_merge(op1.new_outputs, op2,
-          new_outputs)
+          gof.opt.is_same_graph_with_merge(op1.local_outputs, op2,
+          local_outputs)
         - c_code() to remove the double overhead?
-        - opt to unfold it, work inplace on inputs
         - grad() make it support DisconnectedType and the new interface
+        - implement R_op()
         - check how it works with updates.
         - add test with constant as input or inside the inner graph.
         - Add support for the GPU? Probably just need an opt to remove transfer
         - Add support to pickle this Op.
         - Add support/test with random generator
```

Comment:
I didn't replied to this comment. Yes, pre-optimizing the graph before inline could be useful. There is many thread-off point that are sensible to use. This is for later/another PR. I'm not sure what should be the default thread-off point. Here is the one I see. Optimize the inner graph with the optimization phase up to and including:
- useless optimization (remove useless node to make it smaller)
- useless + canonicalize (convert different equivalent graph to a canonical version to help following optimization)
- useless + canonicalize + stabilize opt
- above + specialize (speed opt mostly)

Doing more opt before inline could speed up optimization, in particilar if one OpFromGraph op is reused at many places in the graph. But this could miss some optimization.

Do you have an opinion on what we could do?